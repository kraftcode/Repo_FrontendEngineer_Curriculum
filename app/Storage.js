import localForage from 'localforage';

class Storage {

  constructor (storageService = localForage.createInstance()){ //Default is localForage if no parameter is passed in
    this.storageService = storageService;
  }

  /*
    Persists a key value pair asynchronously using localforage.
    Returns a Promise for the persist operation to the caller.
  */
  asyncPersistAsJSON(key, value) {
    return this.storageService.setItem(key, value);
  }


  /*
    Asynchronously retrieves the corresponding value to the key parameter from storage
    using localforage.
    Returns a Promise for the retrieve operation to the caller.
  */
  asyncRetrieveAsJSON(key) {
    return this.storageService.getItem(key);
  }
}

export default Storage;
