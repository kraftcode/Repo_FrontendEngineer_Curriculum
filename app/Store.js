import moment from 'moment';
import accounting from 'accounting';
import DateFormatter from './lib/dateFormatter';
import Entry from './components/Entry.jsx';
import EntryList from './components/EntryList.jsx';

class Store {

  constructor(_hourlyRate, entryList = {}){
    this.hourlyRate = _hourlyRate;
    this.currentList = entryList;
    this.dateFormatter = new DateFormatter();
  }

  addNewEntry(date = new Date()){
    let _standardStartDate = date;
    let entry = {
      standardStartDate : _standardStartDate,
      startDay : moment(date).format('dd.'),
      startDate : this.dateFormatter.formatDate(_standardStartDate),
      startTime : this.dateFormatter.formatTime(_standardStartDate),
    };
    let id = entry.startDate + entry.startTime;
    if(this.currentList[id]){
      console.error('Store.js addNewEntry(): entryID already exists! ');
    }else {
      this.currentList[id] = entry;
      return id;
    }
  }

  setEndForEntry(entryID, date = new Date()){
    if(this.currentList[entryID]){
      let entry = this.currentList[entryID];
      let standardEndDate = moment(date);
      entry.endDate = this.dateFormatter.formatDate(standardEndDate);
      entry.endTime = this.dateFormatter.formatTime(standardEndDate);
      entry.duration = this.dateFormatter.diffDurationInHours(entry.standardStartDate, standardEndDate);
      entry.earnings = accounting.formatMoney((this.hourlyRate * entry.duration), 'â‚¬', 2, '.', ',');
    } else {
      console.error('Store.js setEndForEntry(): entryID does not exist! ');
      return null;
    }
  }

  removeEntry(entryID){
    delete this.currentList[entryID];
  }

  getStateJSONForStorage(){
    let storageObj = {
      hourlyRate : this.hourlyRate,
      list: this.currentList,
    };
    let key = 0;
    let entryStrings = {};
    entryStrings[key]= storageObj.hourlyRate;
    for(var entry in storageObj.list){
      key++;
      let stringEntry = JSON.stringify(storageObj.list[entry]);
      entryStrings[key] = stringEntry;
    }
    return JSON.stringify(entryStrings);
  }

  getRenderEntryList(){
    let entries = [];
    for(var entry in this.currentList){
      let props = this.currentList[entry];
      if(props.endDate){
        entries.push(<Entry className={props.startDayClass} startDay={props.startDay} startDate={props.startDate} earnings={props.earnings} startTime={props.startTime} endTime={props.endTime} duration={props.duration} />);
      }else{
        entries.push(<Entry className={props.startDayClass} startDay={props.startDay} startDate={props.startDate} earnings={''} startTime={props.startTime} endTime={''} duration={''} />);
      }
    }
    return <EntryList list={entries} />;
  }
}

export default Store;
