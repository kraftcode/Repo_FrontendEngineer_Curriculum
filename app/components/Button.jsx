const Button = (props) => {
  const handleClick = (event) => {
    alert(props.message);
  };
  console.log(props);
  return (
    <button className={props.className} on-click={handleClick}>Click Clack!</button>
  );
};

export default Button;
