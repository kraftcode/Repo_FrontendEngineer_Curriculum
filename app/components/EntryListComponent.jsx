import EntryComponent from './EntryComponent.jsx';

const EntryListComponent = (props) => (
  <ul>
    {props.list.map(function(element){
      return (
        <li>
          <EntryComponent
            startDate={element.startDate}
            endDate={element.endDate}
            earnings={element.earnings}
            duration={element.duration}
          />
        </li>
      );
    })}
  </ul>
);

export default EntryListComponent;
