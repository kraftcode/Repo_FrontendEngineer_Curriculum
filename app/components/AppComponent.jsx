import React from "react";
import { observer } from "mobx-react";
import EntryListComponent from "./EntryListComponent.jsx";
import Button from "./ButtonComponent.jsx";
import LoadingScreen from "./LoadingScreen.jsx";
import TextHeaderComponent from "./TextHeaderComponent.jsx";

const AppComponent = observer(props => {
  const renderComponent = () => {
    if(props.store.isLoading) {
      return <LoadingScreen />;
    } else {
      return(
        <div className="container">
          <TextHeaderComponent company={"Evil Corp."} />
          <div className="buttonfield" />
          <Button active={props.store.active} activeClickHandler={props.store.addNewEntry} inactiveClickHandler={props.store.setEndForEntry} />
          <div className={props.store.active ? "entry__active__topbox" : "entry__inactive__topbox"} />
          <EntryListComponent list={props.store.currentList.reverse()} />
        </div>
      );
    }
  };
  return renderComponent();
});

export default AppComponent;
