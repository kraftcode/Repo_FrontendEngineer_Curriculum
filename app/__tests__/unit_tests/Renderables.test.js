import mobx from "mobx";
import { render, shallow } from "enzyme";
import React from "react";
import Store from "../../Store";
import Storage from "../../Storage";
import AppComponent from "../../components/AppComponent.jsx";
import EntryListComponent from "../../components/EntryListComponent.jsx";
import Button from "../../components/ButtonComponent.jsx";

// Importing and using Store.js as dependency since it was already tested in another test case. (J.F.'s philosophy)

describe.only("Store function tests. ", () => {
  let store;
  let first;
  let second;
  let hourlyTestRate = 42;

  beforeEach(() => {
    store = new Store(new Storage(), hourlyTestRate);
    first = store.addNewEntry(new Date("2015-03-04 15:05:06"));
  });

  it(
    "Should render Loading Message when store is loading from storage. ",
    () => {
      store.setEndForEntry(store.active, new Date("2016-09-09 15:05:06"));
      const tree = render(<AppComponent store={store} />).html();
      expect(tree).toMatchSnapshot();
    }
  );

  it(
    "Should correctly render app tree with inactive button and two entries.",
    () => {
      store.setLoading(false);
      store.setEndForEntry(store.active, new Date("2016-09-09 15:05:06"));
      second = store.addNewEntry(new Date("2016-07-08 15:05:06"));
      store.setEndForEntry(store.active, new Date("2016-09-09 15:05:06"));

      const tree = render(<AppComponent store={store} />).html();
      expect(tree).toMatchSnapshot();
    }
  );

  it("Should render active button component. ", () => {
    const activeButton = render(
      <Button
        active={store.active}
        activeClickHandler={store.addNewEntry}
        inactiveClickHandler={store.setEndForEntry}
      />
    ).html();
    expect(activeButton).toMatchSnapshot();
  });

  it("Should render inactive button component. ", () => {
    store.setEndForEntry(store.active, new Date("2016-09-09 15:05:06"));
    const inactiveButton = render(
      <Button
        active={store.active}
        activeClickHandler={store.addNewEntry}
        inactiveClickHandler={store.setEndForEntry}
      />
    ).html();
    expect(inactiveButton).toMatchSnapshot();
  });

  it("Should render an entry list with two entries. ", () => {
    store.setEndForEntry(store.active, new Date("2016-09-09 15:05:06"));
    store.addNewEntry(new Date("2016-07-08 15:05:06"));
    store.setEndForEntry(store.active, new Date("2016-09-09 15:05:06"));
    const entryList = render(
      <EntryListComponent list={store.getCurrentList().reverse()} />
    ).html();
    expect(entryList).toMatchSnapshot();
  });
});
