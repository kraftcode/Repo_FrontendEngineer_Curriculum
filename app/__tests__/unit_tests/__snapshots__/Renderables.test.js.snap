exports[`Store function tests.  Should render HTML ButtonComponent tag with active button.  1`] = `"<button class=\"button__blue__inactive\">Clock In Now</button>"`;

exports[`Store function tests.  Should render HTML ButtonComponent tag with inactive button.  1`] = `"<button class=\"button__red__active\">Clock Out Now</button>"`;

exports[`Store function tests.  Should render HTML EntryComponent tag built from store list entry without end date.  1`] = `"<div class=\"entry\"><div class=\"entry__top\"><div class=\"entry__startDay\">Mi.</div><div class=\"entry__startDate\">4. März 2015</div><div class=\"entry__earnings\">€0,00</div></div><div class=\"entry__bottom\"><div class=\"entry__startTime\">15:05</div><div class=\"entry__endTime\"><div></div></div><div class=\"entry__duration\"><div></div></div></div></div>"`;

exports[`Store function tests.  Should render HTML EntryComponent tag from store list entry with end date.  1`] = `"<div class=\"entry\"><div class=\"entry__top\"><div class=\"entry__startDay\">Fr.</div><div class=\"entry__startDate\">8. Juli 2016</div><div class=\"entry__earnings\">€63.504,00</div></div><div class=\"entry__bottom\"><div class=\"entry__startTime\">15:05</div><div class=\"entry__endTime\">15:05</div><div class=\"entry__duration\">2:00:00:00</div></div></div>"`;

exports[`Store function tests.  Should render HTML ListEntryComponent tag composed of all elements in store list.  1`] = `"<ul><li> <div class=\"entry\"><div class=\"entry__top\"><div class=\"entry__startDay\">Mi.</div><div class=\"entry__startDate\">4. März 2015</div><div class=\"entry__earnings\">€0,00</div></div><div class=\"entry__bottom\"><div class=\"entry__startTime\">15:05</div><div class=\"entry__endTime\"><div></div></div><div class=\"entry__duration\"><div></div></div></div></div> </li><li> <div class=\"entry\"><div class=\"entry__top\"><div class=\"entry__startDay\">Fr.</div><div class=\"entry__startDate\">8. Juli 2016</div><div class=\"entry__earnings\">€63.504,00</div></div><div class=\"entry__bottom\"><div class=\"entry__startTime\">15:05</div><div class=\"entry__endTime\">15:05</div><div class=\"entry__duration\">2:00:00:00</div></div></div> </li></ul>"`;
